csi/controller_server.go

```diff
@@ -0,0 +1,269 @@
+	}
+}
+
+func (cs *ControllerServer) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (*csi.ListVolumesResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
+	logrus.Infof("ControllerServer create volume req: %v", req)
+	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME); err != nil {
+		logrus.Errorf("CreateVolume: invalid create volume req: %v", req)
+		return nil, err
+	}
+	// Check sanity of request Name, Volume Capabilities
+	if len(req.GetName()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume Name cannot be empty")
+	}
+	if req.GetVolumeCapabilities() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Volume Capabilities cannot be empty")
+	}
+
+	// check for already existing volume name
+	// ID and name are same in longhorn API
+	existVol, err := cs.apiClient.Volume.ById(req.GetName())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if existVol != nil && existVol.Name == req.GetName() {
+		logrus.Debugf("CreateVolume: got an exist volume: %s", existVol.Name)
+		exVolSize, err := util.ConvertSize(existVol.Size)
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		return &csi.CreateVolumeResponse{
+			Volume: &csi.Volume{
+				Id:            existVol.Id,
+				CapacityBytes: exVolSize,
+				Attributes:    req.GetParameters(),
+			},
+		}, nil
+	}
+
+	vol, err := getVolumeOptions(req.GetParameters())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	vol.Name = req.Name
+
+	volSizeBytes := int64(oneGB)
+	if req.GetCapacityRange() != nil {
+		volSizeBytes = int64(req.GetCapacityRange().GetRequiredBytes())
+	}
+	volSizeGB := int(volSizeBytes / 1024 / 1024 / 1024)
+	vol.Size = fmt.Sprintf("%dGi", volSizeGB)
+
+	logrus.Infof("CreateVolume: creating a volume by API client, name: %s, size: %s", vol.Name, vol.Size)
+	resVol, err := cs.apiClient.Volume.Create(vol)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &csi.CreateVolumeResponse{
+		Volume: &csi.Volume{
+			Id:            resVol.Id,
+			CapacityBytes: int64(volSizeBytes),
+			Attributes:    req.GetParameters(),
+		},
+	}, nil
+}
+
+func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
+	logrus.Infof("ControllerServer delete volume req: %v", req)
+	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME); err != nil {
+		logrus.Errorf("DeleteVolume: invalid delete volume req: %v", req)
+		return nil, err
+	}
+
+	existVol, err := cs.apiClient.Volume.ById(req.GetVolumeId())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if existVol == nil {
```

Comment:
The call should be idempotent: https://github.com/container-storage-interface/spec/blob/master/spec.md#deletevolume

Reply:
Done